File: Week09/설진우/PRO_부대복귀.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;

public class Solution {
    public int[] solution(int n, int[][] roads, int[] sources, int destination) {
        int[] answer = new int[sources.length];
        int[] dists = new int[n + 1];
        Arrays.fill(dists, -1);
        boolean[] visited = new boolean[n + 1];
        List<Integer>[] graph = new ArrayList[n + 1];
        for (int i = 0; i <= n; i++) {
            graph[i] = new ArrayList<>();
        }
        for (int[] road : roads) {
            graph[road[0]].add(road[1]);
            graph[road[1]].add(road[0]);
        }

        Queue<Integer> queue = new ArrayDeque<>();
        visited[destination] = true;
        queue.add(destination);
        int dist = 0;
        while (!queue.isEmpty()) {
            List<Integer> tempList = new ArrayList<>();
            while (!queue.isEmpty()) {
                int cur = queue.poll();
                dists[cur] = dist;
                tempList.add(cur);
                visited[cur] = true;
            }

            for (Integer temp : tempList) {
                List<Integer> candidates = graph[temp];
                for (Integer candidate : candidates) {
                    if (!visited[candidate]) {
                        queue.add(candidate);
                    }
                }
            }
            dist++;
        }

        int i = 0;
        for (int source : sources) {
            answer[i] = dists[source];
            i++;
        }

        return answer;
    }
}

File: Week09/김혜진/PRO_부대복귀.js
const n = 5;
const roads = [
  [1, 2],
  [1, 4],
  [2, 4],
  [2, 5],
  [4, 5],
];
sources = [1, 3, 5];
destination = 5;

function solution(n, roads, sources, destination) {
  const graph = Array.from({ length: n + 1 }, () => []);
  for (const [from, to] of roads) {
    graph[from].push(to);
    graph[to].push(from);
  }

  const dist = Array(n + 1).fill(-1);

  const queue = [destination];
  let head = 0;

  dist[destination] = 0;

  while (head < queue.length) {
    const cur = queue[head++];

    for (const next of graph[cur]) {
      if (dist[next] !== -1) continue;

      dist[next] = dist[cur] + 1;
      queue.push(next);
    }
  }

  return sources.map((source) => dist[source]);
}

console.log(solution(n, roads, sources, destination));
